hyper-parameter-optimization.ipynb

```diff
@@ -0,0 +1,180 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import coiled\n",
+    "import dask.distributed"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "cluster = coiled.Cluster(\n",
+    "    n_workers=20, \n",
+    "    configuration=\"coiled/default\", \n",
+    ")\n",
+    "client = dask.distributed.Client(cluster)\n",
+    "\n",
+    "client"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Get and pre-process data"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import dask.dataframe as dd\n",
+    "df = dd.read_csv(\n",
+    "    \"s3://nyc-tlc/trip data/yellow_tripdata_2019-*.csv\", \n",
+    "    parse_dates=[\"tpep_pickup_datetime\", \"tpep_dropoff_datetime\"],\n",
+    "    dtype={\n",
+    "        \"VendorID\": \"UInt8\",\n",
+    "        \"passenger_count\": \"UInt8\",\n",
+    "        \"RatecodeID\": \"UInt8\",\n",
+    "        \"store_and_fwd_flag\": \"category\",\n",
+    "        \"PULocationID\": \"UInt16\",\n",
+    "        \"DOLocationID\": \"UInt16\",    \n",
+    "        \"payment_type\": \"UInt8\",\n",
+    "    },\n",
+    "    blocksize=\"16 MiB\",\n",
+    ")\n",
+    "\n",
+    "data = df[[\"passenger_count\", \"trip_distance\", \"RatecodeID\", \"payment_type\", \"fare_amount\"]]\n",
+    "data = data.fillna(0)\n",
+    "\n",
+    "labels = (df.tip_amount / df.fare_amount) > 0.25\n",
+    "labels = labels.fillna(False)\n",
+    "\n",
+    "from dask_ml.model_selection import train_test_split\n",
+    "\n",
+    "X_train, X_test, y_train, y_test = train_test_split(data, labels, shuffle=True)\n",
+    "X_train, X_test, y_train, y_test = dask.persist(X_train, X_test, y_train, y_test)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Train model"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from sklearn.linear_model import SGDClassifier\n",
+    "from dask_ml.model_selection import HyperbandSearchCV\n",
+    "from scipy.stats import uniform, loguniform\n",
+    "\n",
+    "\n",
+    "clf = SGDClassifier(tol=1e-3, penalty='elasticnet', random_state=0)\n",
+    "\n",
+    "params = {'alpha': loguniform(1e-2, 1e0),  # or np.logspace\n",
+    "          'l1_ratio': uniform(0, 1)}  # or np.linspace\n",
+    "\n",
+    "search = HyperbandSearchCV(clf, params, max_iter=81, random_state=0)\n",
+    "\n",
+    "search.fit(X_train, y_train, classes=[0, 1]);"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Score"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "search.score(X_test.sample(frac=0.1, random_state=123), y_test.sample(frac=0.1, random_state=123))"
```

Comment:
Maybe add a comment on why `frac=0.1` is used?

Reply:
Yeah, this was interesting, and something that we might want to think about in Dask-ML. 

My current understanding is that `search.score` calls a scikit-learn scorer on the inputs, and so these are brought into local memory.  I imagine that this is because we haven't made dask-compatible scorers for everything.  Is that correct?

Reply:
Is [ParallelPostFit][1] relevant?  It takes a trained model and maps the score/predict functions to each chunk.

[1]:https://ml.dask.org/modules/generated/dask_ml.wrappers.ParallelPostFit.html#dask_ml.wrappers.ParallelPostFit

Reply:
Maybe?  (cc @TomAugspurger) 

My sense is that scorers will likely need to be handled one at a time, and that there isn't an obvious way to map them all automatically.  It looks like there is a mapping in `dask_ml/model_selection/scorer.py`.  Maybe ParallelPostFit uses that.  If so, IncrementalSearchCV and friends (Hyperband) should maybe use the same tricks?

Reply:
>  If so, IncrementalSearchCV and friends (Hyperband) should maybe use the same tricks?

Wrapping it in ParallelPostFit should do the trick, but every key in the hyperparameter dict `params` would need to be prepended with `estimator__`.

I'll think more about doing this automatically. My initial reaction is "no", since the default is to fall back to the estimator's default. I wouldn't want to complicate that.

One thing we should be doing is to make something like `Hyperband(..., scoring="accuracy")` work. Right now we use `sklearn.metrics.check_scoring`. But if that used `dask_ml.metrics.check_scoring` things would work. I'll open an issue.

Reply:
> Is ParallelPostFit relevant?

I meant this: `ParallelPostFit(search.best_estimator_).score(X_test, y_test)`.